payments/tests/test_payments_mutations_approve_order.py

```diff
@@ -0,0 +1,214 @@
+
+
+@freeze_time("2020-01-01T08:00:00Z")
+def test_approve_order_does_not_exist(
+    superuser_api_client, payment_provider, notification_template_order_approved,
+):
+    order_id = to_global_id(OrderNode, uuid.uuid4())
+
+    variables = {
+        "orders": [{"orderId": order_id, "email": "foo@bar.com"}],
+    }
+
+    executed = superuser_api_client.execute(APPROVE_ORDER_MUTATION, input=variables)
+    assert len(executed["data"]["approveOrders"]["failedOrders"]) == 1
+    assert executed["data"]["approveOrders"]["failedOrders"][0] == {
+        "id": order_id,
+        "error": "Order matching query does not exist.",
+    }
+
+
+@pytest.mark.parametrize(
+    "order", ["berth_order", "winter_storage_order"], indirect=True,
+)
+@freeze_time("2020-01-01T08:00:00Z")
+def test_approve_order_anymail_error(
+    superuser_api_client,
+    payment_provider,
+    notification_template_order_approved,
+    order: Order,
+):
+    order_id = to_global_id(OrderNode, order.id)
+
+    variables = {
+        "orders": [{"orderId": order_id, "email": "foo@bar.com"}],
+    }
+
+    with patch(
+        "payments.schema.mutations.send_notification",
+        side_effect=AnymailError("Anymail error"),
+    ) as mock:
+        executed = superuser_api_client.execute(APPROVE_ORDER_MUTATION, input=variables)
+    mock.assert_called_once()
+
+    assert len(executed["data"]["approveOrders"]["failedOrders"]) == 1
+    assert executed["data"]["approveOrders"]["failedOrders"][0] == {
+        "id": order_id,
+        "error": "Anymail error",
+    }
+
+
+@pytest.mark.parametrize(
+    "order", ["berth_order", "winter_storage_order"], indirect=True,
+)
+@freeze_time("2020-01-01T08:00:00Z")
+def test_approve_order_one_success_one_failure(
+    superuser_api_client,
+    order: Order,
+    payment_provider,
+    notification_template_order_approved,
+):
+    due_date = (today() + relativedelta(days=14)).date()
+    failure_order_id = to_global_id(OrderNode, uuid.uuid4())
+
+    variables = {
+        "dueDate": due_date,
+        "orders": [
+            {"orderId": failure_order_id, "email": "foo@bar.com"},
+            {
+                "orderId": to_global_id(OrderNode, order.id),
+                "email": order.lease.application.email,
+            },
+        ],
+    }
+    executed = superuser_api_client.execute(APPROVE_ORDER_MUTATION, input=variables)
+    payment_url = payment_provider.get_payment_email_url(
+        order, lang=order.lease.application.language
+    )
+
+    order = Order.objects.get(id=order.id)
+
+    assert len(executed["data"]["approveOrders"]["failedOrders"]) == 1
+    assert executed["data"]["approveOrders"]["failedOrders"][0] == {
+        "id": failure_order_id,
+        "error": "Order matching query does not exist.",
+    }
+
+    assert order.due_date == due_date
+    assert order.lease.status == LeaseStatus.OFFERED
+    assert order.lease.application.status == ApplicationStatus.OFFER_SENT
```

Comment:
Should we check here also that the `failure_order_id ` actually did not get updated?

Reply:
Good idea :Thu

Reply:
On this one we can't test because the order does not exist, but on the previous one yes